Documentation/botr/readytorun-overview.md

```diff
@@ -0,0 +1,335 @@
+    {
+        void intfMethod();
+    }
+    
+    class BaseClass
+    {
+        static int sField;
+        int iField;
+
+        public void iMethod()
+        {
+        }
+
+        public virtual void vMethod(BaseClass aC)
+        {
+        }
+    }
+    
+    class SubClass : BaseClass, Intf
+    {
+        int subField;
+    
+        public override void vMethod(BaseClass aC)
+        {
+        }
+    
+        virtual void intfMethod()
+        {
+        }
+    }
+
+## Instance Field access - LDFLD / STFLD
+
+The CLR stores fields in the 'standard' way, so if RCX holds a BaseClass then
+ 
+    MOV RAX, [RCX + iField_Offset]
+
+will fetch `iField` from this object. `iField_Offset` is a constant known at native code generation time. This is known at compile time only because we mandated that the field layout of `System.Object` is fixed, and thus the entire inheritance chain of `BaseClass` is in the version bubble. It also true even when fields in `BaseClass` contain structs (even from outside the versioning bubble), because we have made it a breaking change to modify the field layout of any public value type. Thus for types whose inheritance hierarchy does not span a version bubble, field fetch is as it always was.
+
+To consider the inter-bubble case, assume that `SubClass` is defined in a different version bubble than BaseClass and we are fetching `subField`. The normal layout rules for classes require `subField` to come after all the files of `BaseClass`. However `BaseClass` could change over time, so we can't wire in a literal constant anymore. Instead we require the following code
+
+	    MOV TMP, [SIZE_OF_BASECLASS]
+	    MOV EAX, [RCX + TMP + subfield_OffsetInSubClass]
+
+	 .data // In the data section
+     SIZE_OF_BASECLASS: UINT32 // One per EXTERN CLASS that is subclassed
+ 
+Which simply assumes that a uint32 sized location has been reserved in the module and that it will be filled in with the size of `BaseClass` before this code is executed. Now a field fetch as one extra instruction, which fetches this size and that dynamic value is used to compute the field. This sequence is a great candidate for CSE (common sub-expression elimination) optimization when multiple fields of the same class are accessed by single method.
+
+A special attention needs to be given to alignment requirements of `SubClass`.
+
+### GC Write Barrier
+
+The .NET GC is generational, which means that most GCs do not collect the whole heap only the 'new' part (which is much more likely to contain garbage. To do this it needs to know the set of roots that point into this 'new' part. This is what the GC write barrier does. Every time an object references that lives in the GC heap is update, bookkeeping code needs to be called to keep log that fact. Any fields whose values were updates are used as potential roots on these partial GCs. The important part here is that any field update of a GC reference must do this extra bookkeeping. 
```

Comment:
- missing closing parens after `likely to contain garbage`

- ~~Every time an object references that lives in the GC heap is update~~ Every time an object *reference* that lives in the GC heap is *updated*

- ~~Any fields whose values were updates~~ Any fields whose values were *updated*